documents/model.go

```diff
@@ -740,17 +749,197 @@ func (m *CoreDocumentModel) ValidateDocumentAccess(docReq *p2ppb.GetDocumentRequ
+// to the read rules.
+func (m *CoreDocumentModel) AddNFT(grantReadAccess bool, registry common.Address, tokenID []byte) (*CoreDocumentModel, error) {
+	data := m.Document.EmbeddedData
+	ndm, err := m.prepareNewVersion(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ndm.Document.EmbeddedData = data
+	cd := ndm.Document
+	nft := getStoredNFT(cd.Nfts, registry.Bytes())
+	if nft == nil {
+		nft = new(coredocumentpb.NFT)
+		// add 12 empty bytes
+		eb := make([]byte, 12, 12)
+		nft.RegistryId = append(registry.Bytes(), eb...)
+		cd.Nfts = append(cd.Nfts, nft)
+	}
+	nft.TokenId = tokenID
+
+	if grantReadAccess {
+		err = ndm.AddNFTToReadRules(registry, tokenID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ndm, ndm.setCoreDocumentSalts()
+}
+
+// IsNFTMinted checks if the there is an NFT that is minted against this document in the given registry.
+func (m *CoreDocumentModel) IsNFTMinted(tokenRegistry TokenRegistry, registry common.Address) bool {
+	nft := getStoredNFT(m.Document.Nfts, registry.Bytes())
+	if nft == nil {
+		return false
+	}
+
+	_, err := tokenRegistry.OwnerOf(registry, nft.TokenId)
+	return err == nil
+}
+
+func getStoredNFT(nfts []*coredocumentpb.NFT, registry []byte) *coredocumentpb.NFT {
+	for _, nft := range nfts {
+		if bytes.Equal(nft.RegistryId[:20], registry) {
+			return nft
+		}
+	}
+
+	return nil
+}
+
+// IsAccountInRole checks if the the account exists in the role provided
+func (m *CoreDocumentModel) IsAccountInRole(roleKey []byte, account identity.CentID) bool {
+	role, err := getRole(roleKey, m.Document.Roles)
+	if err != nil {
+		return false
+	}
+
+	_, found := isAccountInRole(role, account)
+	return found
+}
+
+// GetNFTProofs generate proofs required to mint an NFT
+func (m *CoreDocumentModel) GetNFTProofs(
+	dataRoot []byte,
+	account identity.CentID,
+	registry common.Address,
+	tokenID, roleProof []byte,
+	nftUniqueProof, readAccessProof bool) (proofs []*proofspb.Proof, err error) {
+
+	var pfKeys []string
+	if roleProof != nil {
+		pk, err := getRoleProofKey(m.Document.Roles, roleProof, account)
+		if err != nil {
+			return nil, err
+		}
+
+		pfKeys = append(pfKeys, pk)
+	}
+
+	if nftUniqueProof {
+		pk, err := getNFTUniqueProofKey(m.Document.Nfts, registry)
+		if err != nil {
+			return nil, err
+		}
+
+		pfKeys = append(pfKeys, pk)
+	}
+
+	if readAccessProof {
+		pks, err := getReadAccessProofKeys(m, registry, tokenID)
+		if err != nil {
+			return nil, err
+		}
+
+		pfKeys = append(pfKeys, pks...)
+	}
+
+	signingRootProofHashes, err := m.getSigningRootProofHashes()
+	if err != nil {
+		return nil, errors.New("failed to generate signing root proofs: %v", err)
+	}
+
+	cdtree, err := m.GetDocumentTree()
+	if err != nil {
+		return nil, errors.New("failed to generate document tree: %v", err)
+	}
+
+	// We support fields that belong to different document trees, as we do not prepend a tree prefix to the field, the approach
```

Comment:
This comment is not relevant, since we only create proofs for `cdtree`

Reply:
Ahh. good catch